### DISCUSSION TITLE
How to get srp client host addr with AutoAddress mode

### DISCUSSION THREAD

**Discussion by aaa147147** [openthread/openthread#11059] [Q&A]
I used the AutoHostAddress mode when initializing the SRP client host
`error = otSrpClientEnableAutoHostAddress(esp_openthread_get_instance()); // srp client host address auto`

In my later code, I wanted to obtain the specific IP address of the SRP client host, but I found that I couldn't get it, as shown below
[srp_client.h](https://github.com/openthread/openthread/blob/main/include/openthread/srp_client.h#L72)
```
const otSrpClientHostInfo *srpClientHostInfo = otSrpClientGetHostInfo(instance);
typedef struct otSrpClientHostInfo
{
    const char          *mName;         ///< Host name (label) string (NULL if not yet set).
    const otIp6Address  *mAddresses;    ///< Array of host IPv6 addresses (NULL if not set or auto address is enabled).
    uint8_t              mNumAddresses; ///< Number of IPv6 addresses in `mAddresses` array.
    bool                 mAutoAddress;  ///< Indicates whether auto address mode is enabled or not.
    otSrpClientItemState mState;        ///< Host info state.
} otSrpClientHostInfo;
```
Is there any other way to obtain a specific IP address?

**Comment by aaa147147** [comment]
![image](https://github.com/user-attachments/assets/8176c95a-2c9b-49bf-adf8-5d60d5d25f8a)
I want to obtain this IP for active communication with the SRP server

**Comment by aaa147147** [comment]
If there are no other methods, I suggest that AutoHostAddress mode does not return NULL, but instead returns the actual address used, making it easier for users to use this address

**Reply by wgtdkp** [reply]
> If there are no other methods, I suggest that AutoHostAddress mode does not return NULL, but instead returns the actual address used, making it easier for users to use this address

This makes sense to me, but @abtink I suppose this requires dealing with the ownership of the address buffer as this buffer is not supplied by the client.

**Reply by jwhui** [reply]
Right, the code for selecting the set of auto-registered addresses is here:
https://github.com/openthread/openthread/blob/35608703f75c4c6644988e44b022f4cf131984b2/src/core/net/srp_client.cpp#L1505
It currently relies on iterating through the netif address list.

Ideally, the client should supply the buffer for retrieving the list of addresses.

**Comment by abtink** [comment]
This should be very straightforward.

When "AutoHostAddress" is enabled on an SRP client, it automatically registers all preferred unicast addresses on the Thread network interface, excluding link-local and mesh-local addresses. It will also automatically re-register when addresses on the Thread network interface are updated (e.g., new addresses are added, existing addresses are removed, or marked as non-preferred).

The [`otIp6GetUnicastAddresses()`](https://github.com/openthread/openthread/blob/4e9d362b673bd63e79f5fbfa492ff9959c21fad4/include/openthread/ip6.h#L298-L305) API retrieves the full list of unicast addresses on the Thread network interface. It returns a linked list of [`otNetifAddress`](https://github.com/openthread/openthread/blob/4e9d362b673bd63e79f5fbfa492ff9959c21fad4/include/openthread/ip6.h#L169-L182) entries.

The `mSrpRegistered` flag within an `otNetifAddress` entry indicates if the address is already registered by the SRP client (included in the latest SRP registration message to server).

---

**Clarification on SRP client host addresses:**

The caller can use "AutoHostAddress" or explicitly set the addresses to register using `otSrpClientSetHostAddresses()`, which expects an array of IPv6 addresses as its input parameter. Note that when explicitly set, the provided array must persist (OpenThread SRP Client remembers a pointer to the array and does not make a copy). The `mAddresses` field in the `otSrpClientGetHostInfo` struct then refers to this pointer.

The OpenThread SRP client implementation intentionally does not make copies of strings, text data buffers, addresses, service entries, etc. Instead, it expects the caller to provide buffers for these. This is an intentional design choice to ensure that SRP is fully flexible (e.g., no hard limit on the number of services, addresses, etc.), and it remains memory-efficient. SRP clinet can be used on varierty of device types, including resource-constrained ones.

**Reply by jwhui** [reply]
We may want to clarify the documentation on how to retrieve the list of registered addresses. It is currently buried in `ip6.h` and also states that it's for internal use only.

https://github.com/openthread/openthread/blob/4e9d362b673bd63e79f5fbfa492ff9959c21fad4/include/openthread/ip6.h#L180

Also, "registered addresses" is not the same as "addresses that should be registered". It may be useful to also indicate the set of addresses that the SRP Client is configured to register.

**Reply by abtink** [reply]
It may be useful to add a hint on how to get the full list of addresses and how to interpret the `mSrpRegistered` flag in the SRP client documentation to help those unfamiliar with `otIp6` APIs.

The documentation of `otSrpClientEnableAutoHostAddress` includes the following, which goes into detail about which addresses would be registered under different conditions:

```cpp
/**
 * Enables auto host address mode.
 *
 * When enabled, host IPv6 addresses are automatically set by the SRP client using all preferred unicast addresses on
 * the Thread network interface, excluding link-local and mesh-local addresses. If there is no preferred address, then the Mesh Local
 * EID address is added. The SRP client automatically re-registers when addresses on the Thread network interface are updated
 * (new addresses are added, existing addresses are removed, or are marked as non-preferred).
 *
```
